horovod/torch/__init__.py

```diff
@@ -129,15 +131,43 @@ def broadcast_parameters(params, root_rank):
     `model.named_parameters()`, or `model.parameters()`.
 
     Arguments:
-        params: The list of parameters to broadcast.
+        params: One of the following:
+            - list of parameters to broadcast
+            - dict of parameters to broadcast
+            - torch.optim.Optimizer whose state will be broadcast
         root_rank: The rank of the process from which parameters will be
                    broadcasted to all other processes.
     """
     if isinstance(params, dict):
         params = sorted(params.items())
-    else:
+    elif isinstance(params, list):
         # support both named_parameters() and regular parameters()
         params = [p if isinstance(p, tuple) else (None, p) for p in params]
+    elif isinstance(params, torch.optim.Optimizer):
+        # Newly created optimizers will not have their state initialized, so
+        # do that initialization here
+        if len(params.state_dict()['state']) == 0:
```

Comment:
`params.state_dict()` is actually non-trivial op, we should cache it.  I'd recommend looking `optimizer.load_state_dict()`, `optimizer.state_dict()`.  It seems that there are a lot of corner cases, including CUDA placement.  I wonder if we could just serialize the state_dict, transmit via `hvd.broadcast()`, and load using `optimizer.load_state_dict()`?

Reply:
To make sure I understand correctly, is the suggestion here to serialize the state dict, put it in a tensor, broadcast it with a unique name, then pull out the blob, deserialize it, and load it like so:

```
t = tensor_from_blob(serialize(optimizer.state_dict()))
hvd.broadcast(name, t)
if hvd.rank() > 0:
    state_dict = deserialize(blob_from_tensor(t))
    optimizer.load_state_dict(state_dict)
```

Reply:
If we were to do that, how could we guarantee that the Tensors have the same dimension?

Reply:
Something like this:
```
if hvd.rank == 0:
    t = tensor_from_blob(serialize(optimizer.state_dict()))
    sz = torch.tensor(t.size()).int()
    hvd.broadcast_(name+".sz", sz)
else:
    sz = torch.tensor(3).int() # how many dims?
    hvd.broadcast_(name+".sz", sz)
    t = torch.tensor(sz)
hvd.broadcast_(name, t)
if hvd.rank() > 0:
    state_dict = deserialize(blob_from_tensor(t))
    optimizer.load_state_dict(state_dict)
```